cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,235 @@
+ *
+ * A dictionary column contains a dictionary and a set of indices.
+ * The dictionary is a sorted set of unique values for the column.
+ * The indices values are position indices into the dictionary.
+ */
+class dictionary_column_view : public column_view
+{
+public:
+    dictionary_column_view( column_view dictionary_column );
+    dictionary_column_view( dictionary_column_view&& dictionary_view ) = default;
+    dictionary_column_view( const dictionary_column_view& dictionary_view ) = default;
+    ~dictionary_column_view() = default;
+    dictionary_column_view& operator=(dictionary_column_view const&) = default;
+    dictionary_column_view& operator=(dictionary_column_view&&) = default;
+
+    /**
+     * @brief Returns the internal dictionary column.
+     */
+    column_view dictionary() const noexcept { return _dictionary; }
+
+    /**
+     * @brief Returns the column of indices
+     */
+    column_view indices() const;
+
+private:
+    column_view _dictionary;
+};
+
+namespace dictionary
+{
+
+/**
+ * @brief Create a new dictionary column by adding the new dictionary elements
+ * to the existing dictionary_column.
+ *
+ * The indices are updated if any of the new dictionary elements are sorted
+ * before any of the existing dictionary elements.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = add_dictionary(d1,["b","c"])
+ * d2 is now {["a","b","c","d"],[3,0,2,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary type does not match the dictionary type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to incorporate into the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> add_dictionary( dictionary_column_view const& dictionary_column,
+                                        column_view const& dictionary,
+                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by removing the specified dictionary elements
+ * from the existing dictionary_column.
+ *
+ * The indices are updated to the new positions of the remaining dictionary elements.
+ * Any indices pointing to removed elements are set to null.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = remove_dictionary(d1,["b","c"])
+ * d2 is now {["a","d"],[1,0,null,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary type does not match the dictionary type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to remove from the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> remove_dictionary( dictionary_column_view const& dictionary_column,
+                                           column_view const& dictionary,
```

Comment:
Clearer to call these `keys_to_remove`?

```suggestion
                                           column_view const& keys_to_remove,
```